elastic/fleet-server
# fix unenroll offline agents

## What is the problem this PR solves?

Fix bug with unenrolling offline agents introduced in 8.5

```
"fleet.apikey.id":["49TbhYQBqTAlThoVXNcQ","",""],
"Validation Failed: 1: Field [ids] must not contain blank id, but got blank ids at index positions: [1, 2];\
```

## How does this PR solve the problem?

Added missing null check when creating list of API keys to invalidate.


## How to test this PR locally

- create a Fleet Server policy with enrollment timeout: 60 seconds
- enroll a Fleet Server with docker
- stop the Fleet Server
- start the local Fleet Server and enroll it to kibana with a service token
- wait a few minutes until the unenroll logic kicks in
- expect that the unenroll happens successfully

```
{"log.level":"info","ecs.version":"1.6.0","service.name":"fleet-server","ctx":"policy leader manager","policy_id":"fleet-server-policy","unenroller_uuid":"0114a509-e906-416e-afa9-0bf8f6adaff3","timeout":60000,"fleet.agent.id":"74e84102-000d-4ba7-8e1e-34e9bf70536b","fleet.apikey.id":["49TbhYQBqTAlThoVXNcQ"],"@timestamp":"2022-11-17T14:26:35.334Z","message":"unenrollAgent due to unenroll timeout"}
{"log.level":"info","ecs.version":"1.6.0","service.name":"fleet-server","ctx":"policy leader manager","policy_id":"fleet-server-policy","unenroller_uuid":"0114a509-e906-416e-afa9-0bf8f6adaff3","timeout":60000,"fleet.apikey.id":["74e84102-000d-4ba7-8e1e-34e9bf70536b"],"@timestamp":"2022-11-17T14:26:36.509Z","message":"marked agents unenrolled due to unenroll timeout"}
```

Is this error expected?
 - still seeing these errors continuously coming, found that it is coming from the code here: https://github.com/elastic/fleet-server/blob/main/internal/pkg/dl/agent.go#L82-L84

this change coming from the same pr that caused the original issue: https://github.com/elastic/fleet-server/commit/ca9041fced78f04541fe7dbc7e8495706608355d#diff-81f56287429686d60d33c6df7c9774da997979bc050126d4c807aa602683cc71

```
{"log.level":"error","ecs.version":"1.6.0","service.name":"fleet-server","ctx":"policy leader manager","policy_id":"fleet-server-policy","unenroller_uuid":"0114a509-e906-416e-afa9-0bf8f6adaff3","error.message":"not found","unenroll_timeout":60000,"@timestamp":"2022-11-17T14:27:36.763Z","message":"failed to unenroll offline agents"}
{"log.level":"error","ecs.version":"1.6.0","service.name":"fleet-server","ctx":"policy leader manager","policy_id":"fleet-server-policy","unenroller_uuid":"0114a509-e906-416e-afa9-0bf8f6adaff3","error.message":"not found","unenroll_timeout":60000,"@timestamp":"2022-11-17T14:28:37.016Z","message":"failed to unenroll offline agents"}
```

## Checklist





## Related issues

Closes https://github.com/elastic/fleet-server/issues/2091
